src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,548 @@
+        &upem.to_le_bytes(),
+    );
+    _ = memory.write(
+        &mut caller.as_context_mut(),
+        y_scale as usize,
+        &upem.to_le_bytes(),
+    );
+}
+
+// fn font_get_glyph_extents(font: u32, glyph: u32, extents: *mut CGlyphExtents) -> bool;
+// Returns the glyph's extents for the given glyph ID at current scale and variation settings.
+fn font_get_glyph_extents(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    extents: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut glyph_extents = hb_glyph_extents_t::default();
+    let ret = caller
+        .data()
+        .font
+        .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+    if ret {
+        _ = memory.write(
+            caller.as_context_mut(),
+            extents as usize,
+            bytemuck::bytes_of(&glyph_extents),
+        );
+    }
+
+    ret as u32
+}
+
+// fn font_glyph_to_string(font: u32, glyph: u32, str: *const u8, len: u32);
+// Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+fn font_glyph_to_string(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    str: u32,
+    len: u32,
+) {
+    // len is the assigned heap memory. We should not allocate more than that.
+    // Should not assume we are not writing over anything.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut name = caller
+        .data()
+        .font
+        .glyph_name(GlyphId(glyph as u16))
+        .map(ToOwned::to_owned)
+        .unwrap_or(format!("g{glyph:4}"));
+    name.truncate(len as usize - 1);
+    let name = CString::new(name).unwrap();
+    let name = name.as_bytes();
+
+    _ = memory.write(caller.as_context_mut(), str as usize, name);
+}
+
+// fn font_get_glyph_h_advance(font: u32, glyph: u32) -> i32;
+// Returns the default horizontal advance for the given glyph ID the current scale and variations settings.
+fn font_get_glyph_h_advance(caller: Caller<'_, ShapingData>, _font: u32, glyph: u32) -> i32 {
+    caller.data().font.glyph_h_advance(GlyphId(glyph as u16))
+}
+
+// fn font_get_glyph_v_advance(font: u32, glyph: u32) -> i32;
+// Returns the default vertical advance for the given glyph ID the current scale and variations settings.
+fn font_get_glyph_v_advance(caller: Caller<'_, ShapingData>, _font: u32, glyph: u32) -> i32 {
+    caller.data().font.glyph_v_advance(GlyphId(glyph as u16))
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+enum PointType {
+    MoveTo,
+    LineTo,
+    QuadraticTo,
+    CubicTo,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+struct OutlinePoint {
+    x: f32,
+    y: f32,
+    pointtype: PointType,
+}
+impl OutlinePoint {
+    fn new(x: f32, y: f32, pointtype: PointType) -> Self {
+        OutlinePoint { x, y, pointtype }
+    }
+}
+
+unsafe impl bytemuck::Zeroable for OutlinePoint {}
```

Comment:
Are you sure that the memory layout is the same as harfbuzz expects?

Reply:
I copied it as is from `harfbuzz-wasm`.

Reply:
Ok then.